src/cpp/src/visual_language/cdpruner/kernel_builder.cpp

```diff
@@ -0,0 +1,559 @@
+
+#include "kernel_builder.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <iostream>
+#include <stdexcept>
+
+#include "openvino/op/ops.hpp"
+#include "openvino/op/parameter.hpp"
+#include "openvino/op/result.hpp"
+#include "openvino/openvino.hpp"
+
+namespace ov::genai::cdpruner {
+
+ConditionalKernelBuilder::ConditionalKernelBuilder(const Config& config)
+    : m_config(config),
+      m_requests_initialized(false) {
+    // Precompile models and create infer requests for performance optimization
+    try {
+        ov::Core core;
+
+        // Compile and create infer request for conditional kernel model
+        auto kernel_model = create_conditional_kernel_model();
+        ov::CompiledModel compiled_kernel_model;
+        compiled_kernel_model = core.compile_model(kernel_model, m_config.device);
+        m_conditional_kernel_infer_request = compiled_kernel_model.create_infer_request();
+
+        // Always compile similarity matrix model for potential GPU acceleration
+        auto similarity_model = create_similarity_matrix_model();
+        ov::CompiledModel compiled_similarity_model;
+        compiled_similarity_model = core.compile_model(similarity_model, m_config.device);
+        m_similarity_infer_request = compiled_similarity_model.create_infer_request();
+
+        m_requests_initialized = true;
+
+        if (m_config.pruning_debug_mode) {
+            std::cout << "[CDPruner] ConditionalKernelBuilder: InferRequests initialized for device " << m_config.device
+                      << std::endl;
+        }
+    } catch (const std::exception& e) {
+        if (m_config.pruning_debug_mode) {
+            std::cout << "[CDPruner] ConditionalKernelBuilder: InferRequest initialization failed, will use fallback: "
+                      << e.what() << std::endl;
+        }
+        m_requests_initialized = false;
+    }
+}
+
+ov::Tensor ConditionalKernelBuilder::build(const ov::Tensor& visual_features, const ov::Tensor& input_param) {
+    // Input validation
+    if (visual_features.get_shape().size() != 3) {
+        throw std::invalid_argument("Visual features must be 3D tensor [B, N, D]");
+    }
+
+    auto visual_shape = visual_features.get_shape();
+    size_t batch_size = visual_shape[0];
+    size_t num_tokens = visual_shape[1];
+    size_t feature_dim = visual_shape[2];
+
+    ov::Tensor conditional_kernel;
+    try {
+        conditional_kernel = build_with_ov_model(visual_features, input_param);
+    } catch (const std::exception& e) {
+        if (m_config.pruning_debug_mode) {
+            std::cout << "[CDPruner] ConditionalKernelBuilder: OV model failed, falling back to normal pipeline: "
+                      << e.what() << std::endl;
+        }
+        conditional_kernel = build_with_normal_pipeline(visual_features, input_param);
+    }
+
+    return conditional_kernel;
+}
+
+ov::Tensor ConditionalKernelBuilder::build_with_ov_model(const ov::Tensor& visual_features,
```

Comment:
Can you create helper function for common logic in `build_with_normal_pipeline` and `build_with_normal_pipeline` methods?